File: sandbox/pll_simulation.py
import math

PI = math.pi
PI_2_3 = 2.0 * PI / 3.0
TS = 0.001

class ThreePhase:
    def __init__(self, va=0, vb=0, vc=0):
        self.va = va
        self.vb = vb
        self.vc = vc

class AlphaBeta:
    def __init__(self, alpha=0, beta=0):
        self.alpha = alpha
        self.beta = beta

class DQ:
    def __init__(self, d=0, q=0):
        self.d = d
        self.q = q

class PIController:
    def __init__(self, kp=0, ki=0):
        self.kp = kp
        self.ki = ki
        self.integral = 0.0
        self.error = 0.0
        self.output = 0.0

class PLLState:
    def __init__(self):
        self.theta_est = 0.0
        self.omega_est = 2.0 * PI * 50.0
        self.freq_est = 50.0
        self.pi = PIController(100.0, 1000.0)

def clarke_transform(abc, alphabeta):
    alphabeta.alpha = abc.va
    alphabeta.beta = (abc.vb - abc.vc) / math.sqrt(3.0)

def park_transform(alphabeta, theta, dq):
    cos_theta = math.cos(theta)
    sin_theta = math.sin(theta)
    
    dq.d = alphabeta.alpha * cos_theta + alphabeta.beta * sin_theta
    dq.q = -alphabeta.alpha * sin_theta + alphabeta.beta * cos_theta

def pi_controller_update(pi, error):
    pi.error = error
    pi.integral += error * TS
    pi.output = pi.kp * error + pi.ki * pi.integral

def pll_update(pll, voltage):
    alphabeta = AlphaBeta()
    dq = DQ()
    
    clarke_transform(voltage, alphabeta)
    park_transform(alphabeta, pll.theta_est, dq)
    
    pi_controller_update(pll.pi, dq.q)
    
    pll.omega_est = pll.pi.output
    pll.theta_est += pll.omega_est * TS
    
    while pll.theta_est > 2.0 * PI:
        pll.theta_est -= 2.0 * PI
    while pll.theta_est < 0:
        pll.theta_est += 2.0 * PI
    
    pll.freq_est = pll.omega_est / (2.0 * PI)

def main():
    pll = PLLState()
    
    print("SRF-PLL Test")
    print("Iter\tTheta_in(rad)\tTheta_est(rad)\tFreq_est(Hz)")
    print("---------------------------------------------------")
    
    for i in range(5):
        theta_input = i * PI / 6.0
        
        voltage = ThreePhase(
            math.sin(theta_input),
            math.sin(theta_input - PI_2_3),
            math.sin(theta_input + PI_2_3)
        )
        
        pll_update(pll, voltage)
        
        print(f"{i+1}\t{theta_input:.4f}\t\t{pll.theta_est:.4f}\t\t{pll.freq_est:.2f}")

if __name__ == "__main__":
    main()

File: sandbox/pll_simulation.cs
using System;

namespace PLLSimulation
{
    public class PLL_State
    {
        public double v_alpha;
        public double v_beta;
        public double v_alpha_z1;
        public double v_beta_z1;
        public double vd;
        public double vq;
        public double freq;
        public double freq_hz;
        public double theta;
        public double pi_integral;
        
        public PLL_State()
        {
            const double PI = Math.PI;
            const double GRID_FREQ = 50.0;
            
            v_alpha = 0.0;
            v_beta = 0.0;
            v_alpha_z1 = 0.0;
            v_beta_z1 = 0.0;
            vd = 0.0;
            vq = 0.0;
            freq = 2.0 * PI * GRID_FREQ;
            freq_hz = GRID_FREQ;
            theta = 0.0;
            pi_integral = 0.0;
        }
    }
    
    public class PLLSimulator
    {
        private const double PI = Math.PI;
        private const double SAMPLE_FREQ = 10000.0;
        private const double GRID_FREQ = 50.0;
        private const double SAMPLE_TIME = 1.0 / SAMPLE_FREQ;
        private const double SOGI_K = 0.707;
        private const double PI_KP = 100.0;
        private const double PI_KI = 2000.0;
        
        public static void pll_init(PLL_State pll)
        {
            pll.v_alpha = 0.0;
            pll.v_beta = 0.0;
            pll.v_alpha_z1 = 0.0;
            pll.v_beta_z1 = 0.0;
            pll.vd = 0.0;
            pll.vq = 0.0;
            pll.freq = 2.0 * PI * GRID_FREQ;
            pll.freq_hz = GRID_FREQ;
            pll.theta = 0.0;
            pll.pi_integral = 0.0;
        }
        
        public static void pll_update(PLL_State pll, double v_in)
        {
            double omega = pll.freq;
            double k = SOGI_K;
            double Ts = SAMPLE_TIME;
            
            double v_alpha_new = pll.v_alpha_z1 + k * omega * Ts * (v_in - pll.v_alpha_z1) - omega * Ts * pll.v_beta_z1;
            double v_beta_new = pll.v_beta_z1 + omega * Ts * pll.v_alpha_z1;
            
            pll.v_alpha = v_alpha_new;
            pll.v_beta = v_beta_new;
            pll.v_alpha_z1 = v_alpha_new;
            pll.v_beta_z1 = v_beta_new;
            
            double sin_theta = Math.Sin(pll.theta);
            double cos_theta = Math.Cos(pll.theta);
            
            pll.vd = pll.v_alpha * cos_theta + pll.v_beta * sin_theta;
            pll.vq = -pll.v_alpha * sin_theta + pll.v_beta * cos_theta;
            
            double error = pll.vq;
            
            pll.pi_integral += PI_KI * Ts * error;
            double freq_delta = PI_KP * error + pll.pi_integral;
            
            pll.freq = 2.0 * PI * GRID_FREQ + freq_delta;
            pll.freq_hz = pll.freq / (2.0 * PI);
            
            pll.theta += pll.freq * Ts;
            if (pll.theta > 2.0 * PI)
            {
                pll.theta -= 2.0 * PI;
            }
            if (pll.theta < 0)
            {
                pll.theta += 2.0 * PI;
            }
        }
        
        public static void Main()
        {
            PLL_State pll = new PLL_State();
            pll_init(pll);
            
            Console.WriteLine("Single-phase PLL Algorithm Test - Based on SOGI");
            Console.WriteLine($"Sampling frequency: {SAMPLE_FREQ:F0} Hz, Grid frequency: {GRID_FREQ:F1} Hz");
            Console.WriteLine("Time(s)\tInput Voltage\tEstimated Freq(Hz)\tEstimated Phase(rad)");
            Console.WriteLine(new string('-', 60));
            
            double time = 0.0;
            double final_freq_error = 0.0;
            double final_phase_error = 0.0;
            
            for (int i = 0; i < 1000; i++)
            {
                double v_in = Math.Sin(2.0 * PI * GRID_FREQ * time);
                
                pll_update(pll, v_in);
                
                if (i % 100 == 0)
                {
                    Console.WriteLine($"{time:F4}\t{v_in:F6}\t{pll.freq_hz:F6}\t{pll.theta:F6}");
                }
                
                time += SAMPLE_TIME;
                
                if (i == 999)
                {
                    double expected_phase = (2.0 * PI * GRID_FREQ * time) % (2.0 * PI);
                    final_freq_error = pll.freq_hz - GRID_FREQ;
                    final_phase_error = pll.theta - expected_phase;
                    if (final_phase_error > PI) final_phase_error -= 2.0 * PI;
                    if (final_phase_error < -PI) final_phase_error += 2.0 * PI;
                }
            }
            
            Console.WriteLine("\nFinal Results:");
            Console.WriteLine($"Estimated frequency: {pll.freq_hz:F6} Hz");
            Console.WriteLine($"Frequency error: {final_freq_error:F6} Hz");
            Console.WriteLine($"Phase error: {final_phase_error:F6} rad ({final_phase_error * 180.0 / PI:F6} degrees)");
        }
    }
}